instrumentation/apache-pulsar/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/pulsar/ConsumerImplInstrumentation.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.impl.ConsumerImpl;
+import org.apache.pulsar.client.impl.MessageImpl;
+import org.apache.pulsar.client.impl.PulsarClientImpl;
+
+public class ConsumerImplInstrumentation implements TypeInstrumentation {
+
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    return named("org.apache.pulsar.client.impl.ConsumerImpl");
+  }
+
+  @Override
+  public void transform(TypeTransformer transformer) {
+    String klassName = ConsumerImplInstrumentation.class.getName();
+
+    transformer.applyAdviceToMethod(isConstructor(), klassName + "$ConsumerImplConstructorAdviser");
+
+    transformer.applyAdviceToMethod(
+        isMethod()
+            .and(isProtected())
+            .and(named("messageProcessed"))
+            .and(takesArgument(0, named("org.apache.pulsar.client.api.Message"))),
+        klassName + "$ConsumerImplMethodAdviser");
+  }
+
+  @SuppressWarnings("unused")
+  public static class ConsumerImplConstructorAdviser {
+
+    @Advice.OnMethodExit
+    public static void before(
+        @Advice.This ConsumerImpl<?> consumer,
+        @Advice.Argument(value = 0) PulsarClient client,
+        @Advice.Argument(value = 1) String topic) {
+
+      PulsarClientImpl pulsarClient = (PulsarClientImpl) client;
+      String url = pulsarClient.getLookup().getServiceUrl();
+
+      ClientEnhanceInfo info = new ClientEnhanceInfo(topic, url);
+      ClientEnhanceInfo.virtualField(consumer, info);
+    }
+  }
+
+  @SuppressWarnings("unused")
+  public static class ConsumerImplMethodAdviser {
+
+    @Advice.OnMethodEnter
+    public static Scope before(
+        @Advice.This ConsumerImpl<?> consumer, @Advice.Argument(value = 0) Message<?> message) {
+      ClientEnhanceInfo info = ClientEnhanceInfo.virtualField(consumer);
+      if (null == info) {
+        return Scope.noop();
+      }
+
+      MessageImpl<?> messageImpl = (MessageImpl<?>) message;
+      Context context =
+          PROPAGATOR.extract(Context.current(), messageImpl, MessageTextMapGetter.INSTANCE);
+
+      return TRACER
+          .spanBuilder("Pulsar://ConsumerImpl/messageProcessed")
+          .setParent(context)
+          .setSpanKind(SpanKind.CONSUMER)
+          .setAttribute(TOPIC, info.topic)
+          .setAttribute(SERVICE_URL, info.brokerUrl)
+          .setAttribute(SUBSCRIPTION, consumer.getSubscription())
+          .setAttribute(CONSUMER_NAME, consumer.getConsumerName())
+          .startSpan()
+          .makeCurrent();
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class)
+    public static void after(
+        @Advice.This ConsumerImpl<?> consumer,
+        @Advice.Argument(value = 0) Message<?> message,
+        @Advice.Thrown Throwable t,
+        @Advice.Enter Scope scope) {
+      ClientEnhanceInfo info = ClientEnhanceInfo.virtualField(consumer);
+      if (null == info || scope == null) {
+        if (null != scope) {
+          scope.close();
```

Comment:
In my opinion it would be easier to read if you'd move
```java
if (scope != null) {
  scope.close();
}
```
to the start of the method